homeassistant/components/omnilogic/common.py

```diff
@@ -0,0 +1,157 @@
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    ALL_ITEM_KINDS,
+    ATTR_IDENTIFIERS,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OmniLogicUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching update data from single endpoint."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api: str,
+        name: str,
+        polling_interval: int,
+    ):
+        """Initialize the global Omnilogic data updater."""
+        self.api = api
+
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=name,
+            update_interval=timedelta(seconds=polling_interval),
+        )
+
+    async def _async_update_data(self):
+        """Fetch data from OmniLogic."""
+        try:
+            data = await self.api.get_telemetry_data()
+
+        except OmniLogicException as error:
+            raise UpdateFailed("Error updating from OmniLogic: %s" % error) from error
+
+        parsed_data = {}
+
+        def get_item_data(item, item_kind, current_id, data):
+            """Get data per kind of Omnilogic API item."""
+            if isinstance(item, list):
+                for single_item in item:
+                    data = get_item_data(single_item, item_kind, current_id, data)
+
+            if "systemId" in item:
+                system_id = item["systemId"]
+                current_id = current_id + (item_kind, system_id)
+                data[current_id] = item
+
+            for kind in ALL_ITEM_KINDS:
+                if kind in item:
+                    data = get_item_data(item[kind], kind, current_id, data)
+
+            return data
+
+        parsed_data = get_item_data(data, "Backyard", (), parsed_data)
+
+        return parsed_data
+
+
+class OmniLogicEntity(CoordinatorEntity):
+    """Defines the base OmniLogic entity."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        item_id: tuple,
+        icon: str,
+    ):
+        """Initialize the OmniLogic Entity."""
+        super().__init__(coordinator)
+
+        bow_id = None
+        entity_data = coordinator.data[item_id]
+
+        backyard_id = item_id[:2]
+        if len(item_id) == 6:
+            bow_id = item_id[:4]
+
+        msp_system_id = coordinator.data[backyard_id]["systemId"]
+        entity_friendly_name = f"{coordinator.data[backyard_id]['BackyardName']} "
+        unique_id = f"{msp_system_id}"
+
+        if bow_id is not None:
+            unique_id = f"{unique_id}_{coordinator.data[bow_id]['systemId']}"
+            entity_friendly_name = (
+                f"{entity_friendly_name}{coordinator.data[bow_id]['Name']} "
+            )
+
+        unique_id = f"{unique_id}_{coordinator.data[item_id]['systemId']}_{kind}"
```

Comment:
I'm a bit worried about the unique id. We talked earlier about that the third level device data, `Lights`, `Relays`, `Pumps` etc can change `systemId` if another item in that group is added. So we don't add that id to the unique id.

But then how do we differentiate between one switch entity and another switch entity that represent two different `Relays` that both are of the same kind? There was one example like that in the data you provided.

```json
"Relays": [
  {
    "systemId": "5",
    "relayState": "1",
    "Name": "Ozone",
    "Type": "RLY_HIGH_VOLTAGE_RELAY",
    "Function": "RLY_ACCESSORY",
    "Alarms": []
  },
  {
    "systemId": "14",
    "relayState": "0",
    "Name": "Mister",
    "Type": "RLY_HIGH_VOLTAGE_RELAY",
    "Function": "RLY_ACCESSORY",
    "Alarms": []
  },
  {
    "systemId": "26",
    "relayState": "0",
    "Name": "Water Ft",
    "Type": "RLY_VALVE_ACTUATOR",
    "Function": "RLY_WATER_FEATURE",
    "Alarms": []
  }
],
```

Reply:
Unfortunately I don't think we can solve for that edge case. My expectation is that almost exclusively a new item will receive a new systemId when it is set up on the system, so we shouldn't run up against this one. Because Hayward doesn't use anything else to uniquely identify the entities within their data I think this will have to do. In the unlikely event that this happens, the name would be updated which would trigger a user to note the difference.

Reply:
Ok.